### PR TITLE
Additional LF-integrators

### PR DESCRIPTION
**Pull request by psocratis**
This PR adds the LF integrators (Q,grad V), (Q,curl V) and (Q, div V). A test for the reviewers (or maybe it's a good idea to leave it in the test/convergence) is provided that solves for the Best Approximation Error in the appropriate energy norm.

Note:
For the H(curl) space the 2D and 3D cases are different, since the curl of a 2D vector field is scalar. The two cases are combined into one integrator, where in 2D the integrator expects a FunctionCoefficient and in 3D a VectorFunctionCoefficient.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1460](https://github.com/mfem/mfem/pull/1460) | @psocratis | @tzanio | @mlstowell + @jamiebramwell | 05/19/20 | 06/07/20 | 06/18/20 | |

### CODE REVIEW

File: fem/lininteg.hpp [hunk 1]
```diff
@@ -119,6 +119,33 @@ class DomainLFIntegrator : public DeltaLFIntegrator
    using LinearFormIntegrator::AssembleRHSElementVect;
 };
 
+
+class DomainLFGradIntegrator : public DeltaLFIntegrator
```
Comment (mlstowell) [inline fem/lininteg.hpp:123]: Should probably have a comment like the others

File: tests/convergence/BAE.cpp [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+   {
+      case 0: fec = new H1_FECollection(order,dim);   break;
+      case 1: fec = new ND_FECollection(order,dim);   break;
+      case 2: fec = new RT_FECollection(order-1,dim); break;
+      default: break;
+   }
+   ParFiniteElementSpace *fespace = new ParFiniteElementSpace(pmesh, fec);
+
+   ParGridFunction u_gf(fespace);
+   FunctionCoefficient *u, *divU, *curlU2D;
```
Comment (mlstowell) [inline tests/convergence/BAE.cpp:118]: I don't think these coefficient objects are being deleted anywhere.

File: fem/lininteg.cpp [hunk 3]
```diff
@@ -381,6 +425,127 @@ void VectorFEDomainLFIntegrator::AssembleDeltaElementVect(
...
+   int dof = el.GetDof();
+
+   divshape.SetSize(dof);       // vector of size dof
+   elvect.SetSize(dof);
+   elvect = 0.0;
+
+   const IntegrationRule *ir = IntRule;
+   if (ir == NULL)
+   {
+      // ir = &IntRules.Get(el.GetGeomType(),
```
Comment (mlstowell) [inline fem/lininteg.cpp:523]: Do we need this?

**Comment by mlstowell** [comment]
At the risk of demonstrating my ignorance, what does BAE stand for?

**Comment by psocratis** [comment]
> At the risk of demonstrating my ignorance, what does BAE stand for?

Sorry I was not clear. BAE stands for the Best Approximation Error, i.e., the energy projection. 
Thanks for the review, I'll run it under valgrind on my linux machine.

File: fem/lininteg.hpp (hunk 1 above)
Comment (psocratis) [reply fem/lininteg.hpp:123]: Comment added

File: tests/convergence/BAE.cpp (hunk 2 above)
Comment (psocratis) [reply tests/convergence/BAE.cpp]: Fixed

File: fem/lininteg.cpp (hunk 3 above)
Comment (psocratis) [reply fem/lininteg.cpp]: removed

**Comment by psocratis** [comment]
> I like the BAE test and I'd like to see it in master. It will need some cleanup though. Just a few minor things like an expanded header, comment numbering in the code is inconsistent (abandon the numbers or make them consistent, either would be fine), fix the memory leaks.
> 
> The integrators themselves look great.

cleaned up the BAE example, comments and memory leaks

File: tests/convergence/BAE.cpp [hunk 4]
```diff
@@ -0,0 +1,383 @@
+//
```
Comment (mlstowell) [inline tests/convergence/BAE.cpp:1]: This needs the standard legal mumbo jumbo at the top

File: tests/convergence/BAE.cpp [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+int main(int argc, char *argv[])
+{
+   // 1. Initialize MPI.
+   int num_procs, myid;
+   MPI_Init(&argc, &argv);
+   MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
+   MPI_Comm_rank(MPI_COMM_WORLD, &myid);
+
+   // 2. Parse command-line options.
+   const char *mesh_file = "../data/inline-quad.mesh";
```
Comment (mlstowell) [inline tests/convergence/BAE.cpp:55]: Needs another "../" to find the default mesh file.

Comment (psocratis) [reply tests/convergence/BAE.cpp]: done

File: tests/convergence/BAE.cpp (hunk 4 above)
Comment (psocratis) [reply tests/convergence/BAE.cpp:4]: done

**Comment by jandrej** [comment]
LGTM

**Comment by tzanio** [comment]
@jamiebramwell, can you please review?

File: tests/convergence/BAE.cpp [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+// Compile with: make BAE
+//
+// Sample runs:  mpirun -np 4 BAE -m ../../data/inline-segment.mesh -sr 1 -pr 4 -prob 0 -o 1
```
Comment (jamiebramwell) [inline tests/convergence/BAE.cpp:14]: This is great!

File: tests/convergence/BAE.cpp [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+// Compile with: make BAE
```
Comment (v-dobrev) [inline tests/convergence/BAE.cpp:12]: I'd prefer we stick with lowercase-only filenames: `BAE.cpp` --> `bae.cpp`. For example, `fe.cpp` is lowercase. 😄

Comment (tzanio) [reply tests/convergence/BAE.cpp:12]: In light of https://github.com/mfem/mfem/issues/1254: do we want to commit this?

Comment (psocratis) [reply tests/convergence/BAE.cpp]: > I'd prefer we stick with lowercase-only filenames: `BAE.cpp` --> `bae.cpp`. For example, `fe.cpp` is lowercase. 😄

Sure...I renamed it

Comment (psocratis) [reply tests/convergence/BAE.cpp]: > In light of #1254: do we want to commit this?

I don't know...we kind of forgot about this for a while. Whatever you guys prefer, I am fine with it

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
LGTM

**Comment by tzanio** [comment]
The `test/convergence/bae.cpp` test was moved to #1561.

**Pull request merged**